pallets/asset/src/checkpoint.rs

```diff
@@ -0,0 +1,551 @@
+    /// A series of checkpoints in the future defined by the schedule.
+    pub schedule: CheckpointSchedule,
+    /// The ID of the schedule itself.
+    /// Not to be confused for a checkpoint's ID.
+    pub id: ScheduleId,
+    /// When the checkpoint is next due to be created.
+    /// Used as a cache for more efficient sorting.
+    pub at: Moment,
+}
+
+decl_storage! {
+    trait Store for Module<T: Trait> as Checkpoint {
+        // --------------------- Supply / Balance storage ----------------------
+
+        /// Total supply of the token at the checkpoint.
+        ///
+        /// (ticker, checkpointId) -> total supply at given checkpoint
+        pub TotalSupply get(fn total_supply_at):
+            map hasher(blake2_128_concat) (Ticker, CheckpointId) => T::Balance;
+
+        /// Balance of a DID at a checkpoint.
+        ///
+        /// (ticker, did, checkpoint ID) -> Balance of a DID at a checkpoint
+        pub Balance get(fn balance_at_checkpoint):
+            map hasher(blake2_128_concat) (Ticker, IdentityId, CheckpointId) => T::Balance;
+
+        // ------------------------ Checkpoint storage -------------------------
+
+        /// Checkpoints created per token.
+        ///
+        /// (ticker) -> no. of checkpoints
+        pub Total get(fn total_of):
+            map hasher(blake2_128_concat) Ticker => CheckpointId;
+
+        /// Checkpoints where a DID's balance was updated.
+        /// (ticker, did) -> [checkpoint ID where user balance changed]
+        pub BalanceUpdates get(fn balance_updates):
+            map hasher(blake2_128_concat) (Ticker, IdentityId) => Vec<CheckpointId>;
+
+        /// Checkpoint timestamps.
+        ///
+        /// Every schedule-originated checkpoint maps its ID to its due time.
+        /// Every checkpoint manually created maps its ID to the time of recording.
+        ///
+        /// (checkpoint ID) -> checkpoint timestamp
+        pub Timestamps get(fn timestamps):
+            map hasher(identity) CheckpointId => Moment;
+
+        // -------------------- Checkpoint Schedule storage --------------------
+
+        /// The maximum complexity allowed for an arbitrary ticker's schedule set
+        /// (i.e. `Schedules` storage item below).
+        pub SchedulesMaxComplexity get(fn schedules_max_complexity) config(): u64;
+
+        /// Checkpoint schedule ID sequence for tickers.
+        ///
+        /// (ticker) -> schedule ID
+        pub ScheduleIds get(fn schedule_ids):
+            map hasher(blake2_128_concat) Ticker => ScheduleId;
+
+        /// Checkpoint schedules for tickers.
+        ///
+        /// (ticker) -> [schedule]
+        pub Schedules get(fn schedules):
+            map hasher(blake2_128_concat) Ticker => Vec<StoredSchedule>;
+
+        /// Is the schedule removable?
+        ///
+        /// (ticker, schedule ID) -> removable?
+        pub ScheduleRemovable get(fn schedule_removable):
+            map hasher(blake2_128_concat) (Ticker, ScheduleId) => bool;
```

Comment:
Could this be `double_map`?

Reply:
Querying the removability of all the schedules for a given ticker that have ever existed, but aren't necessarily active, doesn't seem very useful to me. For all the active schedules, you can quickly infer this data by first consulting `Schedules`.